File: cpp/problems/leetcode/0079_word_search/solution.cpp
// LeetCode 79
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    bool exist(vector<vector<char>>& board, string word) {
        int m = (int)board.size(), n = (int)board[0].size();
        function<bool(int, int, int)> dfs = [&](int i, int j, int k) {
            if (k == (int)word.size()) return true;
            if (i < 0 || i >= m || j < 0 || j >= n || board[i][j] != word[k])
                return false;
            char tmp = board[i][j];
            board[i][j] = '#';
            bool ok = dfs(i + 1, j, k + 1) || dfs(i - 1, j, k + 1) ||
                      dfs(i, j + 1, k + 1) || dfs(i, j - 1, k + 1);
            board[i][j] = tmp;
            return ok;
        };
        for (int i = 0; i < m; ++i)
            for (int j = 0; j < n; ++j)
                if (dfs(i, j, 0)) return true;
        return false;
    }
};

int main() {
    vector<vector<char>> b{
        {'A', 'B', 'C', 'E'},
        {'S', 'F', 'C', 'S'},
        {'A', 'D', 'E', 'E'},
    };
    assert(Solution().exist(b, "ABCCED"));
    assert(Solution().exist(b, "SEE"));
    assert(!Solution().exist(b, "ABCB"));
    cout << "0079 OK" << endl;
    return 0;
}

File: python/problems/leetcode/0079_word_search/solution.py
"""LeetCode 79: Word Search."""

from __future__ import annotations


class Solution:
    def exist(self, board: list[list[str]], word: str) -> bool:
        m, n = len(board), len(board[0])

        def dfs(i: int, j: int, k: int) -> bool:
            if k == len(word):
                return True
            if i < 0 or i >= m or j < 0 or j >= n or board[i][j] != word[k]:
                return False
            tmp = board[i][j]
            board[i][j] = "#"
            found = (
                dfs(i + 1, j, k + 1)
                or dfs(i - 1, j, k + 1)
                or dfs(i, j + 1, k + 1)
                or dfs(i, j - 1, k + 1)
            )
            board[i][j] = tmp
            return found

        for i in range(m):
            for j in range(n):
                if dfs(i, j, 0):
                    return True
        return False


if __name__ == "__main__":
    b = [
        ["A", "B", "C", "E"],
        ["S", "F", "C", "S"],
        ["A", "D", "E", "E"],
    ]
    assert Solution().exist(b, "ABCCED")
    assert Solution().exist(b, "SEE")
    assert not Solution().exist(b, "ABCB")
    print("0079 OK")
